ibis/expr/types.py

```diff
@@ -857,18 +886,35 @@ def literal(value, type=None):
     if hasattr(value, 'op') and isinstance(value.op(), ops.Literal):
         return value
 
-    if value is null:
-        dtype = dt.null
+    try:
+        inferred_dtype = dt.infer(value)
+    except com.InputTypeError:
+        has_inferred = False
     else:
-        dtype = dt.infer(value)
+        has_inferred = True
 
-    if type is not None:
+    if type is None:
```

Comment:
Please add some comments here. It's not clear why all of this additional logic was introduced.

Reply:
ok!